# Module 9 – Grouping & Aggregation



## 1. groupby()

`groupby()` is a Pandas method used to divide a DataFrame into groups based on one or more columns.

It helps us organize similar records together and perform calculations such as sum, mean, count, minimum, and maximum on each group.

### Real-World Example

Student records can be grouped based on their class or department to analyze the performance of each group.

### Business Example

A company can group sales data by product category to find the total sales for each category.

### AI/ML Example

During data analysis and preprocessing, `groupby()` can be used to understand patterns and statistics across different categories in a dataset.

In [1]:
import pandas as pd

data = {
    "Name": ["Sai", "Rahul", "Priya", "Amit", "Kiran"],
    "Department": ["IT", "HR", "IT", "HR", "IT"],
    "Marks": [85, 78, 92, 88, 80]
}

df = pd.DataFrame(data)

grouped_marks = df.groupby("Department")["Marks"].mean()

print(grouped_marks)

Department
HR    83.000000
IT    85.666667
Name: Marks, dtype: float64


### Explanation

The `groupby("Department")` groups the students according to their department.

The `mean()` function then calculates the average marks for each department.

The result shows that the average marks for HR are 83.0, while the average marks for IT are approximately 85.67.

### Best Practices

Choose meaningful columns for grouping and apply an appropriate aggregation function to the grouped data.

### Common Mistakes

Grouping by the wrong column or applying an inappropriate aggregation function can produce misleading results.

## 2. agg()

`agg()` is a Pandas method used to apply one or more aggregation functions to grouped data.

It allows us to calculate different statistics such as sum, mean, minimum, maximum, and count.

### Real-World Example

Student marks can be grouped by department and analyzed using average, minimum, and maximum marks.

### Business Example

A company can calculate total sales, average sales, and maximum sales for each product category.

### AI/ML Example

During data analysis, `agg()` can be used to generate multiple summary statistics that help understand different groups in a dataset.

In [2]:
import pandas as pd

data = {
    "Name": ["Sai", "Rahul", "Priya", "Amit", "Kiran"],
    "Department": ["IT", "HR", "IT", "HR", "IT"],
    "Marks": [85, 78, 92, 88, 80]
}

df = pd.DataFrame(data)

result = df.groupby("Department")["Marks"].agg(
    ["mean", "min", "max"]
)

print(result)

                 mean  min  max
Department                     
HR          83.000000   78   88
IT          85.666667   80   92


### Explanation

The `agg()` method applies multiple aggregation functions to the Marks column after grouping the data by Department.

The result shows the average, minimum, and maximum marks for each department.

### Best Practices

Use `agg()` when you need multiple summary statistics from grouped data.

### Common Mistakes

Using an incorrect aggregation function or column can produce incorrect or unexpected results.

## 3. transform()

`transform()` is a Pandas method used to perform calculations on grouped data while keeping the original DataFrame structure.

Unlike `agg()`, which usually returns one summarized row for each group, `transform()` returns a result with the same number of rows as the original data.

### Real-World Example

Student marks can be grouped by department and the average marks of each department can be calculated for every student.

### Business Example

A company can calculate the average sales of each product category and compare every individual sale with its category average.

### AI/ML Example

During data preprocessing, `transform()` can be used to create group-based features while keeping every original record available for further Machine Learning operations.

In [3]:
import pandas as pd

data = {
    "Name": ["Sai", "Rahul", "Priya", "Amit", "Kiran"],
    "Department": ["IT", "HR", "IT", "HR", "IT"],
    "Marks": [85, 78, 92, 88, 80]
}

df = pd.DataFrame(data)

df["Department_Average"] = (
    df.groupby("Department")["Marks"].transform("mean")
)

print(df)

    Name Department  Marks  Department_Average
0    Sai         IT     85           85.666667
1  Rahul         HR     78           83.000000
2  Priya         IT     92           85.666667
3   Amit         HR     88           83.000000
4  Kiran         IT     80           85.666667


### Explanation

The `groupby("Department")` groups the students according to their department.

The `transform("mean")` calculates the average marks for each department and places that average beside every student belonging to that department.

The number of rows remains the same as the original DataFrame.

### Best Practices

Use `transform()` when you need group-level calculations while keeping the original rows.

Use it when the result needs to align with the original DataFrame.

### Common Mistakes

Do not confuse `transform()` with `agg()`. `agg()` produces summarized results, while `transform()` generally returns results aligned with the original rows.

## 4. filter()

`filter()` is a Pandas GroupBy method used to keep or remove entire groups based on a condition.

It checks each group and returns only the groups that satisfy the given condition.

### Real-World Example

Student records can be grouped by department, and we can keep only departments whose average marks are above a certain value.

### Business Example

A company can group sales by product category and keep only categories whose total sales meet a required target.

### AI/ML Example

During data preprocessing, `filter()` can be used to keep only groups that meet specific data quality or analysis conditions.

In [4]:
result = df.groupby("Department").filter(
    lambda group: group["Marks"].mean() > 84
)

print(result)

    Name Department  Marks  Department_Average
0    Sai         IT     85           85.666667
2  Priya         IT     92           85.666667
4  Kiran         IT     80           85.666667


### Explanation

The `filter()` method checks each department group.

The average marks of the IT department are greater than 84, so all IT records are kept.

The HR department has an average of 83, so its records are removed.

### Best Practices

Use `filter()` when the condition needs to be applied to an entire group.

Make sure the filtering condition is clearly defined before removing groups.

### Common Mistakes

Do not confuse `filter()` with row-level filtering. `filter()` can remove or keep an entire group based on the condition.

## 5. Pivot Tables

A Pivot Table is used to summarize and analyze data by grouping values from different columns.

In Pandas, the `pivot_table()` function can calculate summary values such as sum, mean, minimum, and maximum.

### Real-World Example

Student marks can be summarized by department to compare the average performance of different departments.

### Business Example

A company can use Pivot Tables to summarize sales based on product categories and regions.

### AI/ML Example

Pivot Tables can help explore relationships between different variables and understand patterns before Machine Learning.


In [5]:
import pandas as pd

data = {
    "Name": ["Sai", "Rahul", "Priya", "Amit", "Kiran"],
    "Department": ["IT", "HR", "IT", "HR", "IT"],
    "Marks": [85, 78, 92, 88, 80]
}

df = pd.DataFrame(data)

pivot = pd.pivot_table(
    df,
    values="Marks",
    index="Department",
    aggfunc="mean"
)

print(pivot)

                Marks
Department           
HR          83.000000
IT          85.666667


### Explanation

The `pivot_table()` function groups the data based on the Department column.

The `mean` function calculates the average Marks for each department.

The result makes it easy to compare the average marks of different departments.

### Best Practices

Use Pivot Tables when you need a clear summary of data across one or more categories.

Choose the appropriate aggregation function based on the type of analysis.

### Common Mistakes

Using an incorrect column in `values` or choosing an unsuitable aggregation function can produce unexpected results.

## 6. Crosstab

A Crosstab, or Cross-Tabulation, is used to show the relationship between two or more categorical variables.

In Pandas, the `crosstab()` function creates a frequency table that shows how often combinations of values occur.

### Real-World Example

Student data can be analyzed by department and performance category to see how many students belong to each category.

### Business Example

A company can use a Crosstab to compare customers across different locations and product categories.

### AI/ML Example

Crosstab can help analyze relationships between categorical variables and understand the distribution of different classes before Machine Learning.

In [6]:
import pandas as pd

data = {
    "Name": ["Sai", "Rahul", "Priya", "Amit", "Kiran"],
    "Department": ["IT", "HR", "IT", "HR", "IT"],
    "Result": ["Pass", "Pass", "Pass", "Fail", "Pass"]
}

df = pd.DataFrame(data)

result = pd.crosstab(
    df["Department"],
    df["Result"]
)

print(result)

Result      Fail  Pass
Department            
HR             1     1
IT             0     3


### Explanation

The `crosstab()` function counts the combinations of Department and Result.

The result shows that the HR department has 1 student who failed and 1 student who passed.

The IT department has 3 students who passed and no students who failed.

### Best Practices

Use Crosstab when you need to compare the frequency of categorical values.

Choose meaningful categorical columns to make the resulting table useful.

### Common Mistakes

Using numerical columns without first considering whether a frequency table is appropriate can make the result difficult to interpret.

## Advantages

- Makes it easier to summarize large datasets.
- Helps compare different groups in a dataset.
- Provides multiple aggregation functions for analysis.
- Helps identify patterns and relationships between variables.
- Useful for Data Analysis and Machine Learning preprocessing.

## Limitations

- Complex grouping operations can be difficult to understand at first.
- Incorrect grouping can produce misleading results.
- Large datasets may require more memory and processing time.
- Choosing the wrong aggregation function can lead to incorrect analysis.
- Pivot Tables and Crosstabs may become difficult to interpret with many categories.

## Interview Questions and Answers

### 1. What is groupby() in Pandas?

groupby() is used to divide data into groups based on one or more columns and perform calculations on each group.

### 2. What is agg() used for?

agg() is used to apply one or more aggregation functions such as mean, sum, minimum, and maximum.

### 3. What is transform() used for?

transform() performs calculations on grouped data while keeping the same number of rows as the original DataFrame.

### 4. What is the difference between agg() and transform()?

agg() generally returns summarized results, while transform() returns results aligned with the original rows.

### 5. What is filter() used for in GroupBy?

filter() is used to keep or remove entire groups based on a condition.

### 6. What is a Pivot Table?

A Pivot Table summarizes and analyzes data by grouping values from different columns.

### 7. What is pivot_table() used for?

pivot_table() is used to create summarized tables using aggregation functions such as mean, sum, and count.

### 8. What is a Crosstab?

A Crosstab, or Cross-Tabulation, shows the relationship or frequency between categorical variables.

### 9. What is the difference between a Pivot Table and a Crosstab?

A Pivot Table is mainly used for summarized numerical analysis, while a Crosstab is commonly used to show frequencies between categorical variables.

### 10. Why is Grouping & Aggregation important in Data Analysis?

Grouping and aggregation help summarize large datasets, compare different groups, and identify useful patterns in data.

## Self Reflection

I learned how to group and summarize data using Pandas.

I practiced groupby(), agg(), transform(), and filter() for analyzing different groups.

I also learned how to create Pivot Tables and Crosstabs to summarize and compare data.

Overall, I gained a better understanding of how Grouping & Aggregation can be used to analyze datasets and find useful patterns.